tests/Properties.hs

```diff
@@ -535,5 +542,36 @@ misc_tests =
     , testProperty "readIntegerUnsafe" prop_readIntegerUnsafe
     ]
 
+strictness_checks =
+  [ testGroup "Lazy Word8"
+    [ testProperty "foldr is lazy" $ \ xs ->
+        List.genericTake (L.length xs) (L.foldr (:) [ ] (explosiveTail xs)) === L.unpack xs
+    , testProperty "foldr' is strict" $ expectFailure $ \ xs ys ->
+        List.genericTake (L.length xs) (L.foldr' (:) [ ] (explosiveTail (xs <> ys))) === L.unpack xs
+    , testProperty "foldr1 is lazy" $ \ xs -> L.length xs > 0 ==>
+        L.foldr1 const (explosiveTail (xs <> L.singleton 1)) === L.head xs
+    , testProperty "foldr1' is strict" $ expectFailure $ \ xs ys -> L.length xs > 0 ==>
+        L.foldr1' const (explosiveTail (xs <> L.singleton 1 <> ys)) === L.head xs
+    , testProperty "scanl is lazy" $ \ xs ->
+        L.take (L.length xs + 1) (L.scanl (+) 0 (explosiveTail (xs <> L.singleton 1))) === (L.pack . fmap (L.foldr (+) 0) . L.inits) xs
+    , testProperty "scanl1 is lazy" $ \ xs -> L.length xs > 0 ==>
+        L.take (L.length xs) (L.scanl1 (+) (explosiveTail (xs <> L.singleton 1))) === (L.pack . fmap (L.foldr1 (+)) . tail . L.inits) xs
+    ]
```

Comment:
I was wondering why `scanr` is less lazy than `scanl`. The thing is that its output starts from an accumulator, and `Data.ByteString.mapAccumR` is too strict in this respect. 
https://github.com/haskell/bytestring/blob/05a09c3dc66d59e3122dc2fe91392b361d4a6aaa/Data/ByteString.hs#L734-L743

I think this is fine: there are no particular expectations about strictness of `scanr` (there is no `scanr'` in Prelude).

Reply:
Not sure I follow. Is there a specific proposition you are reasoning towards? For example:

**Proposition**   `Data.ByteString.Lazy.scanr` cannot be lazy.

**Proof**

As you noted, the output of a `foldr` starts from the end, so this is the sort of laziness we can have:

```
λ take 2 . reverse $ Prelude.scanr (+) 0 [undefined, 1]
[0,1]
```

So, first the spine of the input list is evaluated to the end, then elements are evaluated from the end backwards. _(Whether the accumulator is evaluated before or after the first element depends on the order of evaluation of `+`.)_ Similarly, the byte stream's spine would have to be evaluated first. But the spine of the byte stream is strict in the leaf:

https://github.com/haskell/bytestring/blob/05a09c3dc66d59e3122dc2fe91392b361d4a6aaa/Data/ByteString/Lazy/Internal.hs#L74-L85

The leaf itself is a byte array and therefore also strict throughout. So, once we force the spine, every byte is also forced. There is no lazy `scanr` for byte streams. ∎

Reply:
This was just a remark for myself and @sjakobi and anyone else who is puzzled why we have laziness properties for `scanl`, but not for `scanr`.

It's not like you cannot make `Data.ByteString.Lazy.scanr` _a bit_ lazier. E. g., for the proposed implementation

```
> Data.ByteString.Lazy.head $ Data.ByteString.Lazy.scanr const 42 ("foo" <> undefined)
*** Exception: Prelude.undefined
CallStack (from HasCallStack):
  error, called at libraries/base/GHC/Err.hs:79:14 in base:GHC.Err
  undefined, called at <interactive>:11:75 in interactive:Ghci1
```

However, if we are ready to sacrifice performance, one can define

```haskell
scanr f z bs = cons hd tl
  where
    (_, tl) = mapAccumR (\x y -> (f y x, x)) z bs
    (hd, _) = List.mapAccumR (\x y -> (f y x, x)) z (unpack bs)
```

for which 

```
> Data.ByteString.Lazy.head $ Data.ByteString.Lazy.scanr const 42 ("foo" <> undefined)
102
```

You can define an even lazier (and slower) version, capable to return first few chunks of bytestring, as long as `f` is very lazy (e. g., `f = const`).

My point is that this is a rare use case, which does not justify performance sacrifices, especially given that there is no general expectation how lazy `scanr` should be. I'm fine with your implementation, no action required.